aws/hll/aws-s3-transfer-manager/src/download/body.rs

```diff
@@ -0,0 +1,223 @@
+            next_seq: 0,
+        }
+    }
+
+    fn push(&mut self, chunk: ChunkResponse) {
+        self.chunks.push(cmp::Reverse(SequencedChunk(chunk)))
+    }
+
+    fn pop(&mut self) -> Option<ChunkResponse> {
+        self.chunks.pop().map(|c| c.0 .0)
+    }
+
+    fn is_ordered(&self) -> bool {
+        let next = self.peek();
+        if next.is_none() {
+            return false;
+        }
+
+        next.unwrap().seq == self.next_seq
+    }
+
+    fn peek(&self) -> Option<&ChunkResponse> {
+        self.chunks.peek().map(|c| &c.0 .0)
+    }
+
+    fn advance(&mut self) {
+        self.next_seq += 1
+    }
+}
+
+#[derive(Debug)]
+struct SequencedChunk(ChunkResponse);
+
+impl Ord for SequencedChunk {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0.seq.cmp(&other.0.seq)
+    }
+}
+
+impl PartialOrd for SequencedChunk {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Eq for SequencedChunk {}
+impl PartialEq for SequencedChunk {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.seq == other.0.seq
+    }
+}
+
+/// A body that returns chunks in whatever order they are received.
+#[derive(Debug)]
+pub(crate) struct UnorderedBody {
+    chunks: Option<mpsc::Receiver<Result<ChunkResponse, TransferError>>>,
+}
+
+impl UnorderedBody {
+    fn new(chunks: Option<BodyChannel>) -> Self {
+        Self { chunks }
+    }
+
+    /// Pull the next chunk of data off the stream.
+    ///
+    /// Returns [None] when there is no more data.
+    /// Chunks returned from an [UnorderedBody] are not guaranteed to be sequenced
+    /// in the right order. Consumers are expected to sequence the data themselves
+    /// using the chunk sequence number (starting from zero).
+    pub(crate) async fn next(&mut self) -> Option<Result<ChunkResponse, TransferError>> {
+        match self.chunks.as_mut() {
+            None => None,
+            Some(ch) => ch.recv().await,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::download::worker::ChunkResponse;
+    use aws_smithy_types::byte_stream::{AggregatedBytes, ByteStream};
+    use bytes::Bytes;
+    use tokio::sync::mpsc;
+
+    use super::{Body, Sequencer};
+
+    fn chunk_resp(seq: u64, data: Option<AggregatedBytes>) -> ChunkResponse {
+        ChunkResponse {
+            seq,
+            data,
+            object_meta: None,
+        }
+    }
+
+    #[test]
+    fn test_sequencer() {
+        let mut sequencer = Sequencer::new();
+        sequencer.push(chunk_resp(1, None));
+        sequencer.push(chunk_resp(2, None));
+        assert_eq!(sequencer.peek().unwrap().seq, 1);
+        sequencer.push(chunk_resp(0, None));
+        assert_eq!(sequencer.pop().unwrap().seq, 0);
+    }
+
+    #[tokio::test]
+    async fn test_body_next() {
+        let (tx, rx) = mpsc::channel(2);
+        let mut body = Body::new(rx);
+        tokio::spawn(async move {
+            for i in 0..3 {
```

Comment:
Can be in a separate test but is it possible we shuffle the order of elements (e.g. [2, 0, 1] or [1, 2, 0] and make sure `body.next()` works as expected? (it may help include the actual order of elements in the diagnostic message of `assert_eq!` in case the test fails and we wish to reproduce the failure)